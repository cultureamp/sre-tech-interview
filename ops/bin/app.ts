#!/usr/bin/env node
import "source-map-support/register";
import * as cdk from "aws-cdk-lib";
import { DockerImageCode, DockerImageFunction } from "aws-cdk-lib/aws-lambda";
import { AttributeType, BillingMode, Table } from "aws-cdk-lib/aws-dynamodb";
import { HttpLambdaIntegration } from "aws-cdk-lib/aws-apigatewayv2-integrations";
import { HttpApi } from "aws-cdk-lib/aws-apigatewayv2";

const app = new cdk.App();
export const stack = new cdk.Stack(app, "comment-vibe");

const code = DockerImageCode.fromImageAsset(`${__dirname}/../..`);
const lambda = new DockerImageFunction(stack, "Lambda", {
  code,
});
const table = new Table(stack, "DynamoTable", {
  tableName: "comment-vibe",
  partitionKey: { name: "surveyId", type: AttributeType.STRING },
  sortKey: { name: "datetime", type: AttributeType.STRING },
  billingMode: BillingMode.PAY_PER_REQUEST,
});
table.grantReadWriteData(lambda);
new HttpApi(stack, "HttpApi", {
  defaultIntegration: new HttpLambdaIntegration("HttpLambda", lambda),
});
