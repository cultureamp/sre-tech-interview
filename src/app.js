const express = require("express");
const aws = require("aws-sdk");

const dynamo = new aws.DynamoDB();
const app = express();

app.get("/", (req, res) => {
  res.json({ message: "OK!" });
});

app.get("/list", async (req, res) => {
  const response = await dynamo
    .scan({
      TableName: process.env["DYNAMODB_TABLE_NAME"],
    })
    .promise();

  res.json({
    action: "list",
    records: response["Items"].map((item) => ({
      id: item.id.S,
      datetime: item.datetime.S,
    })),
    count: response["Count"],
  });
});

app.get("/add/:id", async (req, res) => {
  const record = {
    id: req.params.id,
    datetime: new Date().toISOString(),
  };

  const response = await dynamo
    .putItem({
      TableName: process.env["DYNAMODB_TABLE_NAME"],
      Item: {
        id: { S: record.id },
        datetime: { S: record.datetime },
      },
    })
    .promise();

  res.json({
    action: "add",
    record: record,
  });
});

module.exports = app;
