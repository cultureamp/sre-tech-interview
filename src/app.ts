import * as express from "express";
import * as Sentiment from "sentiment";
import {
  DynamoDBClient,
  ScanCommand,
  PutItemCommand,
} from "@aws-sdk/client-dynamodb";
import { createAWSOptions } from "./aws";

const options = createAWSOptions();
const dynamodb = new DynamoDBClient(options);

const sentiment = new Sentiment();

const createApp = () => {
  const app = express();

  app.use(express.json());

  app.get("/", (req, res) => {
    res.json({ message: "OK!" });
  });

  app.get("/list", async (req, res) => {
    const command = new ScanCommand({
      TableName: "comment-vibe",
    });

    const response = await dynamodb.send(command);
    const items = response["Items"] ?? [];

    res.json({
      action: "list",
      records: items.map((item) => ({
        id: item.id.S,
        content: item.content.S,
        datetime: item.datetime.S,
        sentiment: item.sentiment.N,
      })),
      count: response["Count"],
    });
  });

  app.post("/add", async (req, res) => {
    const analysis = sentiment.analyze(req.body.content);

    const record = {
      id: Date.now(),
      datetime: new Date().toISOString(),
      content: req.body.content,
      sentiment: analysis.score,
    };

    const command = new PutItemCommand({
      TableName: "comment-vibe",
      Item: {
        id: { N: record.id.toString() },
        datetime: { S: record.datetime },
        content: { S: record.content },
        sentiment: { N: record.sentiment.toString() },
      },
    });

    await dynamodb.send(command);

    res.json({
      action: "add",
      record: record,
    });
  });

  return app;
};

export const app = createApp();
