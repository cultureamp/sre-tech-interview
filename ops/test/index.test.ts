import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { stack } from "../bin/app";

test("stack matches snapshot", () => {
  expect(Template.fromStack(stack)).toMatchSnapshot();
});
