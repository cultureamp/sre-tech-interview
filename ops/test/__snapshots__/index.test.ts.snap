// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stack matches snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "DynamoTableB2B22E15": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "surveyId",
            "AttributeType": "S",
          },
          {
            "AttributeName": "datetime",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "surveyId",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "datetime",
            "KeyType": "RANGE",
          },
        ],
        "TableName": "comment-vibe",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "HttpApiDefaultRoute146696CE": {
      "Properties": {
        "ApiId": {
          "Ref": "HttpApiF5A9A8A7",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "$default",
        "Target": {
          "Fn::Join": [
            "",
            [
              "integrations/",
              {
                "Ref": "HttpApiDefaultRouteHttpLambda91177580",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "HttpApiDefaultRouteHttpLambda91177580": {
      "Properties": {
        "ApiId": {
          "Ref": "HttpApiF5A9A8A7",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": {
          "Fn::GetAtt": [
            "LambdaD247545B",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "HttpApiDefaultRouteHttpLambdaPermissionDE2D6043": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "LambdaD247545B",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "HttpApiF5A9A8A7",
              },
              "/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "HttpApiDefaultStage3EEB07D6": {
      "Properties": {
        "ApiId": {
          "Ref": "HttpApiF5A9A8A7",
        },
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
    "HttpApiF5A9A8A7": {
      "Properties": {
        "Name": "HttpApi",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "LambdaD247545B": {
      "DependsOn": [
        "LambdaServiceRoleDefaultPolicyDAE46E21",
        "LambdaServiceRoleA8ED4D3B",
      ],
      "Properties": {
        "Code": {
          "ImageUri": {
            "Fn::Sub": "\${AWS::AccountId}.dkr.ecr.\${AWS::Region}.\${AWS::URLSuffix}/cdk-hnb659fds-container-assets-\${AWS::AccountId}-\${AWS::Region}:42b323473df5cef9755a1b4c9dea32119f45b0069e0925746f442e804e6fc102",
          },
        },
        "PackageType": "Image",
        "Role": {
          "Fn::GetAtt": [
            "LambdaServiceRoleA8ED4D3B",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaServiceRoleA8ED4D3B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaServiceRoleDefaultPolicyDAE46E21": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "DynamoTableB2B22E15",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaServiceRoleDefaultPolicyDAE46E21",
        "Roles": [
          {
            "Ref": "LambdaServiceRoleA8ED4D3B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
